include/sympack/symPACKMatrix2D.hpp

```diff
@@ -865,30 +958,139 @@ namespace symPACK{
+              result = false;
+            }
+          }
+          if (result)
+            logfileptr->OFS()<<"Buffers are equal"<<std::endl;
+          return result;
+        }
+
+#ifdef CUDA_MODE
+        template <typename U=T>
+        void check_close(U * correct, upcxx::global_ptr<U, upcxx::memory_kind::cuda_device> actual, int n) {
+          double epsilon = 1.0;
+          bool iscorrect = true;
+          logfileptr->OFS()<<"BUFFER SIZE: "<<n<<std::endl;
+          U * h_actual = new U[n];
+          CUDA_ERROR_CHECK(cudaMemcpy(h_actual, symPACK::gpu_allocator.local(actual), sizeof(U)*n, cudaMemcpyDeviceToHost));
+          cudaDeviceSynchronize();
+          for (int i=0; i<n; i++) {
+            if (fabs(correct[i] - h_actual[i]) > epsilon && correct[i]!=h_actual[i]) {
+              logfileptr->OFS()<<"Buffers are not equal at index "<<i<<std::endl;
+              logfileptr->OFS()<<"Expected "<<correct[i]<<", got "<<h_actual[i]<<std::endl;
+              logfileptr->OFS()<<"Difference: "<<correct[i]-h_actual[i]<<std::endl;
+              iscorrect = false;
+            }
+          }
+          if (iscorrect)
+            logfileptr->OFS()<<"Buffers are equal"<<std::endl;
+          delete h_actual;
+        }
+
+        //actual must be a device pointer
+        void check_close_dev(T * h_correct, T * actual, int n) {
+          double epsilon = 1.0;
+          bool iscorrect = true;
+          logfileptr->OFS()<<"BUFFER SIZE: "<<n<<std::endl;
+          T * h_actual = new T[n];
+          CUDA_ERROR_CHECK(cudaMemcpy(h_actual, actual, sizeof(T)*n, cudaMemcpyDeviceToHost));
+          for (int i=0; i<n; i++) {
+            if (fabs(h_correct[i] - h_actual[i]) > epsilon && h_correct[i]!=h_actual[i]) {
+              logfileptr->OFS()<<"Buffers are not equal at index "<<i<<std::endl;
+              logfileptr->OFS()<<"Expected "<<h_correct[i]<<", got "<<h_actual[i]<<std::endl;
+              logfileptr->OFS()<<"Difference: "<<h_correct[i]-h_actual[i]<<std::endl;
+              iscorrect = false;
+            }
+          }
+          if (iscorrect)
+            logfileptr->OFS()<<"Buffers are equal"<<std::endl;
+          delete h_actual;
+        }
+#endif
+	
+	void dump_nnz(std::string prelude) {
+		progressptr->OFS()<<prelude<<std::endl;
+		for (int i=0; i<_nnz; i++) {
+			progressptr->OFS()<<_nzval[i]<<",";
+		}
+		progressptr->OFS()<<std::endl;
+	}
+
 
+#ifdef CUDA_MODE
+	bool gpu_alloc_check(std::vector<size_t> bytes) {
+		
+		/* Don't do it if GPU hasn't enough space */
+		size_t free = symPACK::gpu_allocator.segment_size();
+		for (auto const& elem : bytes) {
+			if (elem > free) {
+				return false;
+			}
+		}
+		return true;
+
+	}
+
+	void gpu_alloc_err_handler(size_t alloc_size) {
+		std::cerr<<"ERROR: tried to alloc "<<alloc_size
+			<<" bytes on device with "<<symPACK::gpu_allocator.segment_size()<<" space left"<<std::endl;
+		gdb_lock();
+	}
+#endif
```

Comment:
The two functions above `gpu_alloc_check` and `gpu_alloc_err_handler` use library calls introduced in UPC++ 2023.3.0, and these _might_ be the only use of features introduced in that release. However these functions do not themselves appear to currently have any callers.

It's possible that adding appropriate `#if UPCXX_SPEC_VERSION >= 20230300` here could allow this code to build with older versions of UPC++

Reply:
Yeah neither function is actually used, and they should both probably be removed from the PR. The "GPU error handler" that the code currently uses is really just "do it on the CPU if the GPU allocation failed due to insufficient memory", which does not require a separate function to implement.

Reply:
With those functions removed I suspect the UPC++ floor for ENABLE_CUDA will be 2022.3.0 (for `make_gpu_allocator`), although that's only based on a quick skim and it's possible I missed something.